src/manager/node.c

```diff
@@ -799,31 +897,45 @@ static int node_disconnected(UNUSED sd_bus_message *message, void *userdata, UNU
                 }
 
                 usubs->loaded = false;
+
+                int r = 0;
                 UnitSubscription *usub = NULL;
                 UnitSubscription *next_usub = NULL;
 
                 if (send_state_change) {
                         LIST_FOREACH_SAFE(subs, usub, next_usub, usubs->subs) {
                                 Subscription *sub = usub->sub;
-                                int r = sub->handle_unit_state_changed(
-                                                sub->monitor,
-                                                node->name,
-                                                sub->unit,
-                                                active_state_to_string(usubs->active_state),
-                                                usubs->substate,
-                                                "virtual");
-                                if (r < 0) {
-                                        hirte_log_error("Failed to emit UnitRemoved signal");
+
+                                SubscribedUnit *unit_sub = NULL;
```

Comment:
This seems wrong to me. On subscribe we take the list of subscribed units in the user-specified subscription and for each of the specified units create a new UnitSuibscription and put in the hashtable. So, now that you interate over all the hashtable entries you should get every unit name ever requested. So, there is no need to now iterate over all the units in this subscription again.

In fact, it is very wrong. The send_state_change and the active_state computed above that we are sending *only* applies to the unit we're currently looking at from the hashtable.

Reply:
Good catch! I remove the unnecessary loop.